### PR TITLE
Users-Events

### PR DESCRIPTION
**Pull request by gango-anan**
# I executed this project in 3 parts as explained below:
## Part 1 : Setup and Sign In

- Modeled the data for the application which included the users, events, and attendance tables.
- Initialized a new application and added it to the git version control system.
- Created the database using Postgres.
- Built and migrated the user model.
- Added tests for the user model.
- Added validations for the model along side the tests.
- Created a  Users controller and corresponding routes for #new, #create, and #show actions.
- Created a form for signing up users and signing in users if the users views.
- Created a simple sign-in method that only accepts a username, with the ability to save the user name in a session.

## Part 2 : Basic Events

- Built and migrated the Event model.
- Added associations between the event and the creator of the event.
- Updated the events show page to list all the events that a given user has created.
- Added model tests to ensure the models are running as expected.
- Created the Events controller and the corresponding routes.
- Created a form for creating a new event.
- Updated the create action to aid in the creation of a new event.
- Created the events show page and made it show only the creator of the event.
- Created the events index page and updated it to display all the events.

## Part 3: Event Attendance

- Created and migrated the Attendance model.
- Added association between Event Attendee and Event.
- Created the Attendances Controller and the necessary routes.
- Created a form to aid with adding attendees to the attendance list for a given event.
- Modified the Events show page to display a list of attendees for a specific event.
- Modified the User's show page to display a list of events they are attending(events they are invited to).
- Modified the User's show page to separate events into past and upcoming events categories.
- Modified the Events index page to list all events, separated into Past and Upcoming categories.
- Added styles to all the views.
- Added all the necessary tests.

### CODE REVIEW

File: config/database.yml [hunk 1]
```diff
@@ -20,6 +20,8 @@ default: &default
   # For details on connection pooling, see Rails configuration guide
   # https://guides.rubyonrails.org/configuring.html#database-pooling
   pool: <%= ENV.fetch("RAILS_MAX_THREADS") { 5 } %>
+  username: postgres
+  password: "#Ethel@2416"
```
Comment (oloomoses) [inline config/database.yml]: - [ ] Remove the username and password

File: db/migrate/20210215095939_drop_attendances.rb [hunk 2]
```diff
@@ -0,0 +1,5 @@
+class DropAttendances < ActiveRecord::Migration[6.1]
+  def change
+    drop_table :attendances
+  end
+end
```
Comment (oloomoses) [inline db/migrate/20210215095939_drop_attendances.rb]: - [ ] delete this migration

File: app/views/layouts/_header.html.erb [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+          <%= link_to 'Sign Out', signout_path, class:'nav-link' %>
+        </li>
+      <% else %>
+        <li class="nav-item">
+          <%= link_to 'Sign Up', new_user_path, class:'nav-link' %>
+        </li>
+        <li class="nav-item">
+          <%= link_to 'Sign In', new_login_path, class:'nav-link' %>
+        </li>
+      <% end %>
```
Comment (oloomoses) [inline app/views/layouts/_header.html.erb]: - [ ] Move this logic into helper

File: app/views/users/show.html.erb [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+          </span>
+          <p class="lead"><%= event.description %></p>
+          <hr>
+          <p class="lead"><span class="font-weight-bolder">Venue:</span> <span><%= event.event_venue.capitalize %></span></p>
+          <p class="lead"><span class="font-weight-bolder">
+          Date:</span> <span><%= event.event_date.strftime("%a, %d/%b/%Y, %I:%M %p") %></span></p>
+        </div>
+      </div>
+      <br/>
+    <% end %>
```
Comment (oloomoses) [inline app/views/users/show.html.erb]: - [ ] Move this logic into helpers

**Review by oloomoses — CHANGES_REQUESTED** [review]
# Status: Changes Required
Hello @gango-anan, You have done an awesome job on this project :muscle: :clap: 
## Highlights
- Detailed pull request and easy to understand. Keep it up :100: 
- Well crafted readme :100: 
- Well done on the styling :100: 
## To Improve
- [ ] Kindly check the database.yml settings and remove the login credentials. Not everyone has the same password and username like yours, hence it becomes difficult for local setup `PG::ConnectionBad: FATAL:  Peer authentication failed for user "postgres"`
- [ ] Kindly drop this migration `20210215095939_drop_attendances.rb` it raises error during migration because you created it before the attendances table
- [ ] Some of your rspec tests are failing. Kindly fix them.
- [ ] It would be better to use scopes for the past and upcoming events as stated in the [project requirements](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/scopes_requirement.png)
- [ ] You have logic in your views. Kindly move them to the helpers

Make the above changes before submitting another review.
### Feel free to contact me on slack [Oloo Moses] incase of doubts or concerns
### Oloo Moses

**Review by jubaan — CHANGES_REQUESTED** [review]
[![Microverse][microverse-shield]][microverse-url]
<div>
<h1 align="center"> ⚠️ 🛠️ ⚠️ 🛠️ Changes Required 🛠️ ⚠️ 🛠️ ⚠️ </h1>
<br />
<p><b>"Fail fast, fail often"</b> and don't let this discourage you, because <b>only by failing do we learn to succeed</b>.</p>
<details open="open">
  <summary>Review Summary</summary>
  <ol>
    <li><a href="#congrats"> 💯 Awesome work 👏🏼 👏🏼 </a></li>
    <li><a href="#mandatory"> ⚠️ Things to Improve </a> 🚨 🚨 </li>
    <li><a href="#optional"> 💡Nice to have and other comments ✍🏼 ✍🏼 </a></li>
    <li><a href="#contact"> 🧔🏼 About your Reviewer </a> (you don't know it... but we are just automated bots 🤖) </li>
    <li><a href="#help"> 🆘 <b> 😭 Feeling stuck? or Need help</b>? 🙏🏼 🙏🏼 </a></li>
  </ol>
</details>
</div>

<h3 id="congrats"> 1. :100: Awesome work</h3>
<details close="close">
<summary> [click to open] Big or little achievements, but achievements at the end. <b>Feel proud that you're learning and improving everyday.</b></summary>



</details>

<hr>
<h3 id="mandatory"> 2. ⚠️ Things to Improve (or the To-Do list to get the project approved) </h3> 
<details close="close">
<summary> [click to open] Working on these observations will help you advance and get closer to the goal 🎯, postpone them or do them half-ways and they'll haunt you 👻 👻... FOREVER!! (It will feel like that, trust me). </summary>


- [ ] Not all specs are passing. Please run `rspec` and fix all failing tests.
- [ ] Erase commented code from models


</details>
<p><b>NOTE</b>This list is or could be divided in what is show in this comment and inline comments posted below.<p>
<hr>

<h3 id="optional"> 3. 💡 Nice to have and other comments </h3>
<details close="close">
<summary> [click to open] </h3></summary>


- [ ] Nothing more to add.


</details>

<h3 id="contact"> 4. 🧔🏼 About your Reviewer </h3>
<details close="close">
<summary> [click to open] Say "hello world!" on Slack: @jubaan</summary>
<br />
<div align="center">
  <a src="https://www.github.com/jubaan">
    Julio Añoveros
  </a>
  <p>
    Julio Añoveros | @jubaan</br></br>
    <b>Full-stack Web Developer</b> and</br>
    🤖 Code Reviewer at <a href="https://www.microverse.org/?grsf=kfm2ej">Microverse</a>:rocket:
  </p>
</div>
</details>

<h3 id="help"> 5. 🆘 <b> 😭 Feeling stuck? or Need help</b>? 🙏🏼 🙏🏼</h3>
<details close="close">
  <summary> [click to open] <b>Feel stuck?</b> </summary>
  <br>
  <p> First breath... just do it. Take 15-20 min away from the computer, stretch, meditate, do some exercise, drink water, or do whatever is necessary to distract your mind from the obstacle.</p>
  <p><b>Sometimes you just need to get away to see things with a new perspective.</b></p>
</details>

<details close="close">
  <summary> [click to open] <b>Need help?</b> </summary>
  <br>
  <p> Send me a DM in <a href="https://www.microverse.org/?grsf=kfm2ej">Microverse</a> Students Slack, you'll find me as @jubaan. If I'm not available there are other Code Reviewers you can contact. They'll help as well.</p>
  <p>Remember that <a href="https://www.microverse.org/?grsf=kfm2ej">Microverse</a> has other channels where you can get an answer to your questions:</p>
  <ol>
    <li>The <b>students channel 🧑‍🎓 </b> in slack,</li>
    <li>your <b>teammates 👥 </b>,</li>
    <li>your <b>mentor 👨🏼 👩🏼 </b>, and</li>
    <li><b>Student Success 🏫 </b>.</li>
  </ol>
  <p><b> DON'T hesitate to ask! </b></p>
</details>
<p>If all of the above doesn't help... get a <a href="https://rubberduckdebugging.com/">rubber duck</a> ... it would be great to own one 😄. </p>

<h2 align="center"> Keep up the great work :muscle: and Happy Coding :smiley: :computer:!! </h1>

[microverse-shield]: https://img.shields.io/badge/-microverse-%236F23FF?style=for-the-badge
[microverse-url]: https://www.microverse.org/?grsf=kfm2ej
[bootstrap-shield]: https://img.shields.io/badge/Bootstrap-white?style=for-the-badge&logo=bootstrap
[bootstrap-url]: https://www.microverse.org/?grsf=kfm2ej

**Comment by gango-anan** [comment]
Requested changes made, awaiting approval, thanks.

**Review by jubaan — APPROVED** [review]
[![Microverse][microverse-shield]][microverse-url]
<div>
<h1 align="center"> 
  🎉🥳 🎊 🎉 🥳 🥳 🎊 🎉 🥳🥳 🎊 🎉 🥳 🎉 <br />
  🎉🥳 🎊 🎉 🥳 🥳 🎊 🎉 🥳🥳 🎊 🎉 🥳 🎉 <br />
  🥳 🎊 🎉 🥳  APPROVED 🥳 🎊 🎉 🥳 <br />
  🎉🥳 🎊 🎉 🥳 🥳 🎊 🎉 🥳🥳 🎊 🎉 🥳 🎉 <br />
  🎉🥳 🎊 🎉 🥳 🥳 🎊 🎉 🥳🥳 🎊 🎉 🥳 🎉
</h1>
<h2 align="center">
  What’s done is done. What’s gone is gone. <br />
  One of life’s lessons is always moving on. <br />
  It’s okay to look back to see how far you’ve come but <br /> 
  KEEP MOVING FORWARD.
</h2>

<details open="open">
  <summary>Review Summary</summary>
  <ol>
    <li><a href="#optional"> 💡Nice to have and other comments ✍🏼 ✍🏼 </a></li>
    <li><a href="#contact"> 🧔🏼 About your Reviewer </a> (you don't know it... but we are just automated bots 🤖) </li>
    <li><a href="#help"> 🆘 <b> 😭 Feeling stuck? or Need help</b>? 🙏🏼 🙏🏼 </a></li>
  </ol>
</details>
</div>


<h3 id="optional"> 1. 💡 Nice to have and other comments </h3>
<details close="close">
<summary> [click to open] If you have something here, even when approved, it's just a little push to improve your project.</h3></summary>


- [ ] Nothing more to add.


</details>

<h3 id="contact"> 2. 🧔🏼 About your Reviewer </h3>
<details close="close">
<summary> [click to open] Say "hello world!" on Slack: @jubaan</summary>
<br />
<div align="center">
  <a src="https://www.github.com/jubaan">
    Julio Añoveros
  </a>
  <p>
    Julio Añoveros | @jubaan</br></br>
    <b>Full-stack Web Developer</b> and</br>
    🤖 Code Reviewer at <a href="https://www.microverse.org/?grsf=kfm2ej">Microverse</a>:rocket:
  </p>
</div>
</details>

<h3 id="help"> 3. 🆘 <b> 😭 Feeling stuck? or Need help</b>? 🙏🏼 🙏🏼</h3>
<p> How does it feel? Is nice to advance in the program, right? It keeps you motivated even with all the struggles and takes you one step closer to the goal 🎯!. I wan't you to keep advancing so if you feel stuck and/or need help, remember there is always someone you can reach to keep moving forward, doesn't mater how "dum" you think the questions is.</p>
<details close="close">
  <summary> [click to open] <b>Feel stuck?</b> </summary>
  <br>
  <p> First breath... just do it. Take 15-20 min away from the computer, stretch, meditate, do some exercise, drink water, or do whatever is necessary to distract your mind from the obstacle.</p>
  <p><b>Sometimes you just need to get away to see things with a new perspective.</b></p>
</details>

<details close="close">
  <summary> [click to open] <b>Need help?</b> </summary>
  <br>
  <p> Send me a DM in <a href="https://www.microverse.org/?grsf=kfm2ej">Microverse</a> Students Slack, you'll find me as @jubaan. If I'm not available there are other Code Reviewers you can contact. They'll help as well.</p>
  <p>Remember that <a href="https://www.microverse.org/?grsf=kfm2ej">Microverse</a> has other channels where you can get an answer to your questions:</p>
  <ol>
    <li>The <b>students channel 🧑‍🎓 </b> in slack,</li>
    <li>your <b>teammates 👥 </b>,</li>
    <li>your <b>mentor 👨🏼 👩🏼 </b>, and</li>
    <li><b>Student Success 🏫 </b>.</li>
  </ol>
  <p><b> DON'T hesitate to ask! </b></p>
</details>
<p>If all of the above doesn't help... get a <a href="https://rubberduckdebugging.com/">rubber duck</a> ... it would be great to own one 😄. </p>

<h1 align="center"> See you a the top 🚀!! </h1>

[microverse-shield]: https://img.shields.io/badge/-microverse-%236F23FF?style=for-the-badge
[microverse-url]: https://www.microverse.org/?grsf=kfm2ej
[bootstrap-shield]: https://img.shields.io/badge/Bootstrap-white?style=for-the-badge&logo=bootstrap
[bootstrap-url]: https://www.microverse.org/?grsf=kfm2ej

**Pull request merged**